raft/raft.go

```diff
@@ -632,7 +632,7 @@ func stepLeader(r *raft, m pb.Message) {
 			r.id, r.raftLog.lastTerm(), r.raftLog.lastIndex(), r.Vote, m.From, m.LogTerm, m.Index, r.Term)
 		r.send(pb.Message{To: m.From, Type: pb.MsgVoteResp, Reject: true})
 	case pb.MsgSnapStatus:
-		if pr.State != ProgressStateSnapshot {
+		if pr == nil || pr.State != ProgressStateSnapshot {
```

Comment:
`MsgBeat` and `MsgCheckQuorum` are internally-generated messages with `m.From == r.id`. If `r.id` is not in `r.prs`, I think it should be fine to drop those messages too. Then `MsgVote` is the only message that a leader may need to handle from a node not in `r.prs`. 

`MsgVote` handling is split between the top-level `Step` method and the per-state `step{Leader,Candidate,Follower}`. If we could move the sending of rejections from `stepLeader` to `Step`, we could do the `r.prs` check at the start of `stepLeader` no matter what the message type is.